Loconet throttle setSpeedSetting() options

SteveRawlinson opened the pull request:
A suggestion on jmri-developers of making available an option for users to choose to allow 'duplicate' loconet slot speed messages met with some scepticism.

Since that option is only really useful for people controlling trains with scripts rather than (physical or on-screen) throttles, this PR takes a different approach: allow scripts to call any of three different methods on Loconet throttles:

- setSpeedSetting(float) - does exactly what it has always done
- setSpeedSettingAgain(float) - sends a loconet message even when the speed is the same as the existing speed
- setSpeedSetting(float, boolean, boolean) - allow or suppress 'duplicate' loconet messages according to options specified

Hopefully this approach avoids the potential pitfalls of the previous suggestion: extra support queries, unintentional additional loconet traffic.

SteveRawlinson commented:
Are there any systems other loconet which suppress speed messages if the new speed is effectively equal to the existing speed? I looked and couldn't see any but I might have missed something.

pabender commented:
> On Jan 24, 2018, at 5:21 PM, Steve Rawlinson <notifications@github.com> wrote:
> Are there any systems other loconet which suppress speed messages if the new speed is effectively equal to the existing speed? I looked and couldn't see any but I might have missed something.
> 
Whether there are or not today is irrelevant to my comment.

If any system finds that to be a useful feature to implement in the future, having the interface in place allows that.

This also allows us to modify example scripts we distribute, which typically are not Loconet specific, so they use the new calls where appropriate.

SteveRawlinson commented:
> Whether there are or not today is irrelevant to my comment.

I understand why you'd prefer changes put into the interface, that makes perfect sense. But it would be odd to do that and then not provide an implementation for underlying systems which swallow duplicate speed messages. I can do the loconet implementation, obviously,  but I might not be able to do it for anything else because I can't test it. 

I checked to see if I could find any others and I couldn't. If there are no others I can make the changes you've requested in about 10 minutes. If there are others I'd have to have a rethink.

pabender commented:
Sent from my iPad

> On Jan 25, 2018, at 4:07 AM, Steve Rawlinson <notifications@github.com> wrote:
> I understand why you'd prefer changes put into the interface, that makes perfect sense. But it would be odd to do that and then not provide an implementation for underlying systems which swallow duplicate speed messages. I can do the loconet implementation, obviously, but I might not be able to do it for anything else because I can't test it.
> 
My comment was not JUST to add this to the interface.  You also need add a default implementation in the Abstract Throttle that calls the existing setSpeedSetting method.  You don't need to worry about any logic to make it function beyond that.

For testing that, duplicate the testSetSpeedSetting method in AbstractThrottleTest for each of the new methods and modify the duplicate to test the appropriate method.  The testing infrastructure will make sure this runs against all throttle implementations.

( the Loconet throttle test can then override that test to check for system specific behavior... )

SteveRawlinson commented:
On 25/01/2018 13:31, Paul Bender wrote:
> My comment was not JUST to add this to the interface. You also need add a default implementation in the Abstract Throttle that calls the existing setSpeedSetting method. 

Yes, I understand that, no problem.

What I want to know is whether there are any current systems  - other than loconet - that should have these options actually do something? I can't find any but I want to check.

SteveRawlinson commented:
Added new methods and options to the interface and AbstractThrottle. Added a test. Please don't merge, the test is failing and I need to work out why.

SteveRawlinson commented:
Contrary to my last comment, the tests are passing. The changes requested by @pabender are implemented (if I've understood them correctly). The travis/appveyor failures are not, as far as I can see, relevant to this PR.

Before merging I'd appreciate confirmation that loconet is the only existing underlying throttle implementation that could be expected to use these options.

rhwood commented:
@pabender Is this now acceptable?
